File: src/Ithkuil/Allomorph.hs
{-# LANGUAGE OverloadedStrings #-}
-- | Ithkuil V4 Ca Complex: Construction, Parsing, and Allomorphic Rules
--
-- The Ca complex (Slot VI) encodes Configuration, Extension, Affiliation,
-- Perspective, and Essence as a single consonant cluster. It is built
-- compositionally from 4 component tables (Ca1-Ca4), then allomorphic
-- substitution rules are applied.
--
-- For parsing, all 3840 possible Ca forms are pre-generated and stored
-- in a reverse lookup map.
module Ithkuil.Allomorph
  ( constructCa
  , parseCaSlot
  , renderCa
  , caForward
  , caReverse
  , geminateCa
  , degeminateCa
  , isGeminateCa
  ) where

import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Text (Text)
import qualified Data.Text as T
import Ithkuil.Grammar

--------------------------------------------------------------------------------
-- Component Tables
--------------------------------------------------------------------------------

-- | Ca1: Configuration consonant (from grammar ch03 table)
ca1 :: Configuration -> Text
ca1 UNI = ""
ca1 DPX = "s"
ca1 DSS = "c"
ca1 DSC = "ks"
ca1 DSF = "ps"
ca1 DDS = "ţs"
ca1 DDC = "fs"
ca1 DDF = "š"
ca1 DFS = "č"
ca1 DFC = "kš"
ca1 DFF = "pš"
ca1 MSS = "t"
ca1 MSC = "k"
ca1 MSF = "p"
ca1 MDS = "ţ"
ca1 MDC = "f"
ca1 MDF = "ç"
ca1 MFS = "z"
ca1 MFC = "ž"
ca1 MFF = "ẓ"

-- | Ca2: Extension consonant (non-UNIPLEX, voiceless form)
-- Placed between Configuration and Perspective in the Ca complex
ca2 :: Extension -> Text
ca2 DEL = ""
ca2 PRX = "t"
ca2 ICP = "k"
ca2 ATV = "p"
ca2 GRA = "g"
ca2 DPL = "b"

-- | Ca2 standalone: Extension consonant for UNIPLEX Configuration
-- These are voiced forms used when Config is UNI (no Ca1 consonant)
ca2Standalone :: Extension -> Text
ca2Standalone DEL = ""
ca2Standalone PRX = "d"
ca2Standalone ICP = "g"
ca2Standalone ATV = "b"
ca2Standalone GRA = "gz"
ca2Standalone DPL = "bz"

-- | Ca3: Affiliation prefix
-- (standalone form, before-consonant form)
-- Standalone = UNI/DEL with no other components present
-- Before-consonant = prefix before Configuration consonant
ca3 :: Affiliation -> Text
ca3 CSL = ""
ca3 ASO = "l"
ca3 COA = "r"
ca3 VAR = "ř"

-- | Ca3 standalone: Affiliation when used alone (UNI/DEL/M/NRM + affiliation)
ca3Standalone :: Affiliation -> Text
ca3Standalone CSL = ""
ca3Standalone ASO = "nļ"
ca3Standalone COA = "rļ"
ca3Standalone VAR = "ň"

-- | Ca4: Perspective + Essence consonant
-- (standalone form, after-consonant form)
-- Standalone = when Ca1, Ca2, Ca3 are all empty
-- After-consonant = when any of Ca1, Ca2, Ca3 is present
ca4 :: Perspective -> Essence -> (Text, Text)
ca4 M_ NRM = ("l", "")
ca4 G_ NRM = ("r", "r")
ca4 N_ NRM = ("v", "w")
ca4 A_ NRM = ("j", "y")
ca4 M_ RPV = ("tļ", "l")
ca4 G_ RPV = ("ř", "ř")
ca4 N_ RPV = ("m", "m")
ca4 A_ RPV = ("n", "n")

--------------------------------------------------------------------------------
-- Ca Construction
--------------------------------------------------------------------------------

-- | Construct raw Ca from components (before allomorphic substitutions)
-- Grammar order: Affiliation + Configuration + Extension + Perspective/Essence
-- Special cases:
-- 1. UNIPLEX with Extension → use ca2Standalone (voiced forms: d, g, b, gz, bz)
-- 2. UNIPLEX with Affiliation only → use ca3Standalone (nļ, rļ, ň)
-- 3. Fully standalone (UNI/CSL/DEL/M_/NRM) → standalone perspective form
-- 4. General: Aff-prefix + Config + Extension + Perspective suffix
constructCaRaw :: SlotVI -> Text
constructCaRaw (co, af, pe, ex, es)
  -- UNIPLEX with Extension (voiced standalone forms)
  | co == UNI && ex /= DEL =
      ca2Standalone ex <> perspSuffix pe es
  -- UNIPLEX with only Affiliation (standalone forms)
  | co == UNI && af /= CSL =
      ca3Standalone af <> perspSuffix pe es
  -- Fully standalone (UNI/CSL/DEL) → perspective standalone form
  | co == UNI =
      fst (ca4 pe es)
  -- General: Affiliation prefix + Configuration + Extension + Perspective
  | otherwise = affPfx <> cfg <> ext <> persp''
  where
    affPfx = ca3 af
    cfg = ca1 co
    ext = ca2 ex
    -- Perspective + Essence (after-consonant form since config is present)
    persp = snd (ca4 pe es)
    -- Special combination rules from grammar section 3.5.1:
    -- N_ RPV (m) → h when preceded by [C]t, [C]k, or [C]p
    -- A_ RPV (n) → ç when preceded by [C]t, [C]k, or [C]p
    persp''
      | persp == "m" && endsWithStop = "h"
      | persp == "n" && endsWithStop = "ç"
      | otherwise = persp
    preceding = affPfx <> cfg <> ext
    endsWithStop = case T.unsnoc preceding of
      Just (_, c) -> c `elem` ("tkp" :: [Char])
      Nothing -> False
    -- Perspective suffix for UNIPLEX with Extension/Affiliation
    perspSuffix p e = snd (ca4 p e)

-- | Construct Ca with allomorphic substitutions applied
constructCa :: SlotVI -> Text
constructCa = applySubstitutions . constructCaRaw

--------------------------------------------------------------------------------
-- Allomorphic Substitutions
--------------------------------------------------------------------------------

-- | Allomorphic substitution rules for Ca consonant clusters
-- From morphology v1.3.1 Sec 3.6 table (page 14)
-- Simple substitutions: always apply
simpleSubstitutions :: [(Text, Text)]
simpleSubstitutions =
  [ ("pp", "mp"), ("tt", "nt"), ("kk", "nk")
  , ("pb", "mb"), ("kg", "ng")
  , ("ll", "pļ"), ("rr", "ns")
  , ("çy", "nd"), ("řř", "nš")
  ]

-- | Context-dependent substitutions: [C]X → [C]Y
-- Only apply when the pattern occurs after at least one consonant (not word-initial)
contextSubstitutions :: [(Text, Text)]
contextSubstitutions =
  [ ("gm", "x"), ("gn", "ň")
  , ("bm", "v"), ("bn", "ḑ")
  , ("çx", "xw")  -- [C]çx → [C]xw
  ]

-- | Second-pass substitutions for multi-step derivations
-- Applied after context substitutions produce intermediate forms
-- fbm → [C]bm→[C]v → fv → vw
-- ţbn → [C]bn→[C]ḑ → tḑ → ḑy
secondPassSubstitutions :: [(Text, Text)]
secondPassSubstitutions =
  [ ("fv", "vw"), ("tḑ", "ḑy")
  ]

-- | Apply all substitution rules in order:
-- 1. Simple substitutions (apply everywhere)
-- 2. Context-dependent substitutions (only non-initial)
-- 3. Second-pass for multi-step derivations
applySubstitutions :: Text -> Text
applySubstitutions = applySimple secondPassSubstitutions
                   . applyContext contextSubstitutions
                   . applySimple simpleSubstitutions
  where
    applySimple [] t = t
    applySimple ((from, to):rest) t = applySimple rest (T.replace from to t)
    -- Replace pattern only when it occurs after at least one character
    applyContext [] t = t
    applyContext ((from, to):rest) t = applyContext rest (replaceNonInitial from to t)

-- | Replace occurrences of a pattern in text, but only when not at position 0
replaceNonInitial :: Text -> Text -> Text -> Text
replaceNonInitial from to t
  | T.length t < T.length from = t
  | otherwise =
    let len = T.length from
        -- Keep the first character, then do replacements in the rest
        -- But we need to check at each position after index 0
        go i
          | i + len > T.length t = t  -- no more room for pattern
          | T.take len (T.drop i t) == from && i > 0 =
              T.take i t <> to <> replaceNonInitial from to (T.drop (i + len) t)
          | otherwise = go (i + 1)
    in go 0

--------------------------------------------------------------------------------
-- Pre-generated Lookup Tables
--------------------------------------------------------------------------------

-- | All possible SlotVI values
allSlotVI :: [SlotVI]
allSlotVI = (,,,,) <$> allOf <*> allOf <*> allOf <*> allOf <*> allOf

-- | Forward map: SlotVI -> Ca consonant cluster
caForward :: Map SlotVI Text
caForward = Map.fromList [(s, constructCa s) | s <- allSlotVI]

-- | Reverse map: Ca consonant cluster -> SlotVI
-- When multiple SlotVI produce the same Ca form, the first one wins
caReverse :: Map Text SlotVI
caReverse = Map.fromList [(constructCa s, s) | s <- reverse allSlotVI]

-- | Parse a Ca consonant cluster to SlotVI
parseCaSlot :: Text -> Maybe SlotVI
parseCaSlot = flip Map.lookup caReverse

-- | Render a SlotVI to Ca consonant cluster
renderCa :: SlotVI -> Text
renderCa s = Map.findWithDefault (constructCa s) s caForward

--------------------------------------------------------------------------------
-- Ca Gemination (Sec 3.6.1)
--------------------------------------------------------------------------------

-- | Geminate a Ca consonant cluster (Sec 3.6.1 rules).
-- Uses pre-generated map for lookup; falls back to algorithmic rules.
geminateCa :: Text -> Text
geminateCa ca = Map.findWithDefault (geminateCaForm ca) ca caGemForward

-- | Degeminate a consonant cluster back to its Ca form.
-- Returns the input unchanged if not a recognized geminated form.
degeminateCa :: Text -> Text
degeminateCa t = Map.findWithDefault t t caGemReverse

-- | Check if a consonant cluster is a geminated Ca form.
isGeminateCa :: Text -> Bool
isGeminateCa = flip Map.member caGemReverse

-- | Forward gemination map: Ca form → geminated form
caGemForward :: Map Text Text
caGemForward = Map.fromList
  [(ca, geminateCaForm ca) | ca <- Map.keys caReverse]

-- | Reverse gemination map: geminated form → Ca form
caGemReverse :: Map Text Text
caGemReverse = Map.fromList
  [(geminateCaForm ca, ca) | ca <- Map.keys caReverse]

-- Character classification for gemination rules

isCaStop :: Char -> Bool
isCaStop c = c `elem` ("tkpdgb" :: [Char])

isVoicelessStop :: Char -> Bool
isVoicelessStop c = c `elem` ("tkp" :: [Char])

isLiquidApprox :: Char -> Bool
isLiquidApprox c = c `elem` ("lrřwy" :: [Char])

isSibilant :: Char -> Bool
isSibilant c = c `elem` ("sšzžçcč" :: [Char])

isNonSibFric :: Char -> Bool
isNonSibFric c = c `elem` ("fţvḑ" :: [Char])

isCaNasal :: Char -> Bool
isCaNasal c = c `elem` ("nmň" :: [Char])

-- | Fricatives for rule 6 (voiceless stop + fricative)
isFricR6 :: Char -> Bool
isFricR6 c = c `elem` ("sšfţç" :: [Char])

isLRŘ :: Char -> Bool
isLRŘ c = c `elem` ("lrř" :: [Char])

-- | Find index of first sibilant in text
findFirstSibilant :: Text -> Maybe Int
findFirstSibilant = go 0 . T.unpack
  where
    go _ [] = Nothing
    go i (c:cs)
      | isSibilant c = Just i
      | otherwise = go (i+1) cs

-- | Duplicate the character at position i: "kst" at 1 → "ksst"
gemCharAt :: Int -> Text -> Text
gemCharAt i t = T.take (i+1) t <> T.drop i t

-- | Rule 7: Two stops at end → special substitution
gemStopStop :: [(Text, Text)]
gemStopStop =
  [ ("pt", "bbḑ"), ("pk", "bbv")
  , ("kt", "ggḑ"), ("kp", "ggv")
  , ("tk", "ḑvv"), ("tp", "ddv")
  ]

-- | Rule 8: Stop + nasal at end → special substitution
gemStopNasal :: [(Text, Text)]
gemStopNasal =
  -- 8a: voiceless stop + nasal
  [ ("pm", "vvm"), ("pn", "vvn")
  , ("km", "xxm"), ("kn", "xxn")
  , ("tm", "ḑḑm"), ("tn", "ḑḑn")
  -- 8b: voiced stop + nasal
  , ("bm", "mmw"), ("bn", "mml")
  , ("gm", "ňňw"), ("gn", "ňňl")
  , ("dm", "nnw"), ("dn", "nnl")
  ]

-- | Apply gemination rules 1-8 (without rule 9 l/r/ř prefix handling)
geminateCaInner :: Text -> Text
geminateCaInner t
  | T.null t = t
  | T.length t == 1 = t <> t                              -- Rule 1
  | t == "tļ" = "ttļ"                                      -- Rule 2
  | isCaStop c1 && isLiquidApprox c2 = T.cons c1 t        -- Rule 3
  | Just i <- findFirstSibilant t = gemCharAt i t          -- Rule 4
  | isNonSibFric c1 || isCaNasal c1 = T.cons c1 t         -- Rule 5
  | isVoicelessStop c1 && isFricR6 c2 = gemCharAt 1 t     -- Rule 6
  | Just gem <- lookup t gemStopStop = gem                 -- Rule 7
  | Just gem <- lookup t gemStopNasal = gem                -- Rule 8
  | otherwise = T.cons c1 t                                -- Fallback
  where
    c1 = T.head t
    c2 = T.index t 1

-- | Apply gemination to a Ca form (Sec 3.6.1, rules 1-9)
geminateCaForm :: Text -> Text
geminateCaForm t
  | T.null t = t
  | T.length t == 1 = t <> t                              -- Rule 1
  -- Rule 9: l-/r-/ř- prefix → apply rules to rest
  | isLRŘ c1 && T.length t > 1 =
      let inner = T.tail t
          gemInner = geminateCaInner inner
      in if "çç" `T.isInfixOf` gemInner || "ļļ" `T.isInfixOf` gemInner
         then T.cons c1 (T.cons c1 (T.tail t))  -- fallback: geminate prefix
         else T.cons c1 gemInner
  -- Rules 1-8
  | otherwise = geminateCaInner t
  where c1 = T.head t

File: allomorph/allomorph.go
// Package allomorph builds the Slot VI Ca consonant cluster from its
// five components (Configuration, Affiliation, Perspective, Extension,
// Essence). The raw composition is in ConstructCaRaw; allomorphic
// substitutions and the bidirectional lookup tables live alongside it.
package allomorph

import (
	g "github.com/christian-oudard/ithkuil/grammar"
)

// ca1 is the Configuration component of Ca (grammar ch.3 table).
// UNI is empty — no consonant is contributed.
var ca1 = [...]string{
	g.UNI: "",
	g.DPX: "s",
	g.DSS: "c",
	g.DSC: "ks",
	g.DSF: "ps",
	g.DDS: "ţs",
	g.DDC: "fs",
	g.DDF: "š",
	g.DFS: "č",
	g.DFC: "kš",
	g.DFF: "pš",
	g.MSS: "t",
	g.MSC: "k",
	g.MSF: "p",
	g.MDS: "ţ",
	g.MDC: "f",
	g.MDF: "ç",
	g.MFS: "z",
	g.MFC: "ž",
	g.MFF: "ẓ",
}

// ca2 is the Extension component in the voiceless form, used when a
// Configuration consonant precedes it.
var ca2 = [...]string{
	g.DEL: "",
	g.PRX: "t",
	g.ICP: "k",
	g.ATV: "p",
	g.GRA: "g",
	g.DPL: "b",
}

// ca2Standalone is the Extension component in the voiced form, used when
// Configuration is UNI (no Ca1 consonant to follow).
var ca2Standalone = [...]string{
	g.DEL: "",
	g.PRX: "d",
	g.ICP: "g",
	g.ATV: "b",
	g.GRA: "gz",
	g.DPL: "bz",
}

// ca3 is the short Affiliation prefix (used before a Configuration
// consonant).
var ca3 = [...]string{
	g.CSL: "",
	g.ASO: "l",
	g.COA: "r",
	g.VAR: "ř",
}

// ca3Standalone is the long Affiliation form, used when Affiliation is
// the only marked component (no Configuration, no Extension, no
// perspective suffix).
var ca3Standalone = [...]string{
	g.CSL: "",
	g.ASO: "nļ",
	g.COA: "rļ",
	g.VAR: "ň",
}

// ca4Entry pairs the standalone and after-consonant forms of a
// Perspective+Essence combination. Standalone is used when Ca1, Ca2, Ca3
// are all empty; the after-consonant form is used otherwise.
type ca4Entry struct{ standalone, suffix string }

var ca4Table = map[ca4Key]ca4Entry{
	{g.M_, g.NRM}: {"l", ""},
	{g.G_, g.NRM}: {"r", "r"},
	{g.N_, g.NRM}: {"v", "w"},
	{g.A_, g.NRM}: {"j", "y"},
	{g.M_, g.RPV}: {"tļ", "l"},
	{g.G_, g.RPV}: {"ř", "ř"},
	{g.N_, g.RPV}: {"m", "m"},
	{g.A_, g.RPV}: {"n", "n"},
}

type ca4Key struct {
	Perspective g.Perspective
	Essence     g.Essence
}

func ca4(p g.Perspective, e g.Essence) ca4Entry {
	return ca4Table[ca4Key{p, e}]
}

// ConstructCaRaw builds the raw Ca consonant cluster from a SlotVI by
// concatenating its components in grammatical order
// (Affiliation + Configuration + Extension + Perspective/Essence) with
// special-case forms for UNI Configuration. Allomorphic substitutions
// are NOT applied here — see ApplySubstitutions for that next step.
//
// Special-case rules (grammar §3.5):
//
//  1. UNI + Extension ≠ DEL: use voiced standalone Extension form
//     (d/g/b/gz/bz) followed by the perspective suffix.
//  2. UNI + Affiliation ≠ CSL: use the long Affiliation form alone if
//     no perspective suffix; otherwise short prefix + suffix.
//  3. UNI/CSL/DEL (fully default Configuration/Affiliation/Extension):
//     use the standalone perspective form.
//  4. General: Affiliation prefix + Configuration + Extension +
//     Perspective suffix, with grammar §3.5.1 N_/A_ RPV → h/ç when
//     preceded by a stop consonant.
func ConstructCaRaw(s g.SlotVI) string {
	persp := ca4(s.Perspective, s.Essence)

	if s.Configuration == g.UNI && s.Extension != g.DEL {
		return ca2Standalone[s.Extension] + persp.suffix
	}
	if s.Configuration == g.UNI && s.Affiliation != g.CSL {
		if persp.suffix == "" {
			return ca3Standalone[s.Affiliation]
		}
		return ca3[s.Affiliation] + persp.suffix
	}
	if s.Configuration == g.UNI {
		return persp.standalone
	}

	// General compositional form.
	preceding := ca3[s.Affiliation] + ca1[s.Configuration] + ca2[s.Extension]
	suffix := persp.suffix
	if suffix == "m" && endsWithStop(preceding) {
		suffix = "h"
	} else if suffix == "n" && endsWithStop(preceding) {
		suffix = "ç"
	}
	return preceding + suffix
}

func endsWithStop(s string) bool {
	if s == "" {
		return false
	}
	r := []rune(s)
	last := r[len(r)-1]
	return last == 't' || last == 'k' || last == 'p'
}

// init verifies that the component tables cover every value in their
// respective enums. Running this at startup catches table omissions
// immediately instead of returning empty strings at parse time.
func init() {
	for _, c := range g.AllConfigurations {
		if int(c) >= len(ca1) {
			panic("allomorph: ca1 table missing entry for " + c.String())
		}
	}
	for _, e := range g.AllExtensions {
		if int(e) >= len(ca2) || int(e) >= len(ca2Standalone) {
			panic("allomorph: ca2 table missing entry for " + e.String())
		}
	}
	for _, a := range g.AllAffiliations {
		if int(a) >= len(ca3) || int(a) >= len(ca3Standalone) {
			panic("allomorph: ca3 table missing entry for " + a.String())
		}
	}
	for _, p := range g.AllPerspectives {
		for _, e := range g.AllEssences {
			if _, ok := ca4Table[ca4Key{p, e}]; !ok {
				panic("allomorph: ca4 table missing " + p.String() + "/" + e.String())
			}
		}
	}
	// Sanity check the default Ca encodes to "l".
	if got := ConstructCaRaw(g.DefaultSlotVI); got != "l" {
		panic("allomorph: default Ca raw should be \"l\", got " + got)
	}
}
